msal/http_decorate.py

```diff
@@ -0,0 +1,114 @@
+from threading import Lock
+from hashlib import sha256
+
+from .individual_cache import _IndividualCache as IndividualCache
+from .individual_cache import _ExpiringMapping as ExpiringMapping
+
+
+# https://datatracker.ietf.org/doc/html/rfc8628#section-3.4
+DEVICE_AUTH_GRANT = "urn:ietf:params:oauth:grant-type:device_code"
+
+
+def _hash(raw):
+    return sha256(repr(raw).encode("utf-8")).hexdigest()
+
+
+def _handle_http_429_5xx_retry_after(result=None, **ignored):
+    assert result
```

Comment:
I would be surprised by the `AssertionError` because the default value in the signature tells me `None` is an acceptable value

Reply:
Thanks for catching this kind of details!

Admittedly, it looks weird at first glance. The signature here seemingly contains `None` as the default value, only because the actual callback function's shape is decided by the `_IndividualCache`'s `__call__()`. In that shape, the signature is defined as a bunch of position-less parameters, and `result` is not even the first parameter. So, here in this layer, we use `def _handle_http_429_5xx_retry_after(result=None, **ignored):` to match that shape.

The `assert` here would never be triggered in real run, however, I deliberately add it to remind maintainers that the `result` should/would not be `None`.

Not sure how much of those info I can squeeze in a short sentence in my next commit. :-)